added a basic web frontend

stephens2424 opened the pull request:
I wanted an easy way to check up on the fuzzer running from my server, so I wrote up this little frontend. I kept it pretty simple for the time being. Let me know if you'd like me to change anything to merge, but I thought you might like this as part of the project.

dvyukov commented:
Thanks for working on this. I am a bit busy this week. I will take a look later this week hopefully.

dgryski commented on go-fuzz/main.go:
```diff
@@ -13,6 +16,8 @@ import (
 	"syscall"
 	"time"
+
+	"github.com/stephens2424/writerset"
```
This should point to the vendored writerset implementation.

dgryski commented on go-fuzz/main.go:
```diff
@@ -29,10 +34,13 @@ var (
 	flagCoverCounters = flag.Bool("covercounters", true, "use coverage hit counters")
 	flagSonar         = flag.Bool("sonar", true, "use sonar hints")
 	flagV             = flag.Int("v", 0, "verbosity level")
+	flagMasterStats   = flag.String("masterstats", "", "serve master stats handlers on that address")
```
"masterstats HTTP server listen address"

dgryski commented on go-fuzz/main.go:
```diff
@@ -88,5 +99,31 @@ func main() {
+func broadcastStats(stats masterStats) {
+	b, err := json.Marshal(stats)
+	if err != nil {
+		panic(err)
```
`panic()` seems excessive here.

stephens2424 replied:
It's a fairly simple struct type and marshaling it should *never* cause an actual error unless somebody writes in a related bug, so a panic seemed to make sense. The only sensical alternative I could see is logging the error and skip writing to the output.

dgryski commented on go-fuzz/stats.go:
```diff
@@ -0,0 +1,142 @@
+
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
+<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/js/bootstrap.min.js"></script>
+<script src="../../assets/js/ie10-viewport-bug-workaround.js"></script>
```
The ie10 workaround should be loaded via a CDN or included in the repository.

dgryski commented on go-fuzz/main.go:
```diff
@@ -88,5 +99,31 @@ func main() {
+			io.WriteString(w, statsPage)
+		})
+
+		go http.ListenAndServe(*flagMasterStats, evMux)
```
The error return from `http.ListenAndServe()` is silently discarded.

dgryski commented:
The README should also be updated.  Since the statistics page isn't served on `/`, I had to read the source code to find out what page to load in my browser.

stephens2424 commented:
I changed the statistics page to just serve on `/`.

I also updated the `http.ListenAndServe` calls to both handle errors. Previously, I just wrote mine to match the pprof implementation, but the errors should probably be checked for in both cases.

I can also rebase/squash before merge, but thought the finer grained commit would be helpful for now.

dgryski commented:
LGTM, but I don't have a commit-bit on this repo.  Waiting for @dvyukov to review/merge.

dgryski commented:
I wonder if it makes sense to have the web front-end and the pprof endpoint both use the same listen address?  So, attach the masterstats endpoint to the default muxer and call `http.ListenAndServe()` if either the stats end point or the pprof endpoint were requested.

stephens2424 commented:
The options I can see related to that:

- Keep both flags as strings, and if the addresses happen to be the same, just silently collapse them to the same listener.
- Add a new flag, `-http`, which specifies the address to listen on, and convert `-pprof` and `-masterstats` to be boolean. The program would fail to start if the combination of flags doesn't make sense (e.g. specifying `-http` but not either of the specific to serve)

dvyukov commented on go-fuzz/internal/writerset/writerset.go:
```diff
@@ -0,0 +1,98 @@
+// writers are easy to add or remove as necessary.
+type WriterSet struct {
+	m  map[io.Writer]chan error
+	mu *sync.Mutex
```
This should be just sync.Mutex, not a pointer.

stephens2424 replied:
Can you educate me as to why?

dvyukov replied:
Because storing by value is the default. It is faster, consumes less memory, generates less garbage, less code (no need to initialize) and removes questions like "is this thing shared? does anybody else holds a pointer to it?". You use pointers only when there is a need to use a pointer. E.g. you want to allocate and initialize something lazily, or you want to share an object (two objects hold a pointer to the same mutex). Here is no need to use a pointer here.

dvyukov commented on go-fuzz/internal/writerset/writerset.go:
```diff
@@ -0,0 +1,98 @@
+	ws.mu.Lock()
+	defer ws.mu.Unlock()
+
+	c, ok := ws.m[w]
```
We never add the same writer twice, so remove this part (or replace with panic).

stephens2424 replied:
On a lot of the ones about writerset, I'm going to have this same comment: I wrote that package to be a tad more general than the use cases here and it's hosted as an independent library at github.com/stephens2424/writerset, but I've vendored it here (manually). So yes, parts of this aren't used in this project, but I think it's fitting for them to stay around. Of all the possible approaches to this, I think some type of vendoring makes sense, but if you'd rather pare down this to just what's used here, we can do that, but I'd like to double check before going ahead.

dvyukov replied:
okay

but comments about mutex and Write method is something that you may want to fix upstream

stephens2424 replied:
absolutely, they were good comments. I just pushed some changes there. working on the full changes to push up here

dvyukov commented on go-fuzz/internal/writerset/writerset.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// Contains determines if w is in the set.
+func (ws *WriterSet) Contains(w io.Writer) bool {
```
remove this method, it is not used

dvyukov commented on go-fuzz/internal/writerset/writerset.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// Remove ensures w is not in the set.
+func (ws *WriterSet) Remove(w io.Writer) {
```
remove this method, it is not used

dvyukov commented on go-fuzz/internal/writerset/writerset.go:
```diff
@@ -0,0 +1,98 @@
+// that writer is removed from the set. The error will be sent on the channel created when
+// the writer was added.
+func (ws *WriterSet) Write(b []byte) (int, error) {
+	var err error
```
this err is never assigned, in the loop you create a new err variable
so just return nil always

dvyukov commented on go-fuzz/internal/writerset/writerset.go:
```diff
@@ -0,0 +1,98 @@
+	}
+
+	if n == 0 {
+		return 0, err
```
This and the following return statement do not look correct to me.

Caller should not be affected by presence/absence of registered writers.
However, when writers are present you return some non-zero amount of bytes; while when there are no writers you return 0. What should caller do in that case? Retry write? If there are no registered writers, we still have successfully written len(b) bytes to zero underlying writers.
Also that average number of bytes looks weird. I don't see how it can be ever useful.
Just return (len(b), nil) unconditionally.

dvyukov commented on go-fuzz/main.go:
```diff
@@ -40,8 +48,16 @@ func main() {
 	if *flagPprof != "" {
-		go http.ListenAndServe(*flagPprof, nil)
+		go func() {
+			err := http.ListenAndServe(*flagPprof, nil)
```
panic(http.ListenAndServe(*flagPprof, nil))
ListenAndServe does not return with nil error

dvyukov commented on go-fuzz/main.go:
```diff
@@ -29,10 +34,13 @@ var (
 	flagCoverCounters = flag.Bool("covercounters", true, "use coverage hit counters")
 	flagSonar         = flag.Bool("sonar", true, "use sonar hints")
 	flagV             = flag.Int("v", 0, "verbosity level")
+	flagMasterStats   = flag.String("masterstats", "", "masterstats HTTP server listen address")
```
please call the flag just "http"
I want to expose more stats via http, including hub/slave stats. Http is also shorter and used by godoc and go tool trace.
But the check that this flag must not be provided on slave is correct -- all stats should be aggregated and exposed by master.

dvyukov commented on go-fuzz/main.go:
```diff
@@ -88,5 +104,36 @@ func main() {
 		go slaveMain()
 	}
 
+	if *flagMasterStats != "" {
```
I feel that these handlers should be defined on Master and moved to master.go, along with writerSet and broadcastStats.
That would reduce number of global variables, and make it easier to export more stats (e.g. if there is a new crasher, you probably want to look at output and download the file remotely as well).

dvyukov commented on go-fuzz/master.go:
```diff
@@ -71,25 +71,58 @@ func masterLoop(m *Master) {
+		Corpus:           uint64(len(m.corpus.m)),
+		Crashers:         uint64(len(m.crashers.m)),
+		StartTime:        m.startTime,
+		LastMutationTime: m.lastInput,
```
s/LastMutationTime/LastNewInputTime/
mutations happen all the time, this is the time when we generated new interesting input and added it to corpus

dvyukov commented on go-fuzz/stats.go:
```diff
@@ -0,0 +1,167 @@
+var evtSource = new EventSource("/eventsource");
+evtSource.addEventListener("ping", function(e) {
+	var data = JSON.parse(e.data);
+	var uptime = formatDuration(Date.now() - Date.parse(data.StartTime))
```
data.StartTime is time on server running go-fuzz
Date.now() is local time on client machine.
The difference is meaningless.
Also, I don't want to maintain long javascript functions.
Let's just push formatted uptime from server (we already have format function there).
If we ever need raw uptime in seconds and/or server start time/current time, we can always add them later.

dvyukov commented:
Regarding -http, -pprof and -masterstats.
I think we can leave just -http and enable pprof and stats handlers always. I don't see any downsides of this.
I would even enable http serving on some default port, and leave -http flag only to override the port.

stephens2424 commented:
A couple questions above. I might not have a chance to implement changes to the PR for a week or so (I'm traveling abroad). Thanks for the feedback!

stephens2424 commented:
I also inlined the assets. I realized the way it was written, it requires an internet connection to run the binary, which I think is suboptimal.

dvyukov commented on go-fuzz/master.go:
```diff
@@ -2,15 +2,21 @@ package main
 	"sync/atomic"
 	"time"
+
+	"github.com/stephens2424/writerset"
```
this should be go-fuzz/internal/writerset

dvyukov commented on go-fuzz/master.go:
```diff
@@ -27,6 +33,8 @@ type Master struct {
 	statRestarts  uint64
 	coverFullness int
+
+	writerSet *writerset.WriterSet
```
please rename this var to something more descriptive, e.g. "statWriters" or "statWriterSet"

dvyukov commented:
besides these two comments looks good to me

stephens2424 commented:
Need anything else here?

dvyukov commented:
Didn't notice that you uploaded a new version. Sorry.
Merging

dvyukov commented:
Thank you for working on this!

stephens2424 commented:
My pleasure! Thanks for accepting and helping improve it!

Thomasdezeeuw commented:
Are there any docs about this new feature (besides `go-fuzz -h`)?

Also we opening the webbrowser it displays the folder contents by default, a redirect to stats.html or even better displaying stats.html at index (url: `/`) would be nice.
